omp/solver/upper_trs_kernels.cpp

```diff
@@ -97,26 +101,35 @@ void solve(std::shared_ptr<const OmpExecutor> exec,
     auto row_ptrs = matrix->get_const_row_ptrs();
     auto col_idxs = matrix->get_const_col_idxs();
     auto vals = matrix->get_const_values();
+    const auto num_rows = matrix->get_size()[0];
+    const auto num_rhs = b->get_size()[1];
+    dense::fill(exec, x, nan<ValueType>());
 
-#pragma omp parallel for
-    for (size_type j = 0; j < b->get_size()[1]; ++j) {
-        for (size_type inv_row = 0; inv_row < matrix->get_size()[0];
-             ++inv_row) {
+#pragma omp parallel for schedule(monotonic : auto) collapse(2)
+    for (size_type inv_row = 0; inv_row < matrix->get_size()[0]; ++inv_row) {
+        for (size_type j = 0; j < b->get_size()[1]; ++j) {
             auto row = matrix->get_size()[0] - 1 - inv_row;
             auto diag = one<ValueType>();
-            x->at(row, j) = b->at(row, j);
+            auto val = b->at(row, j);
             for (auto k = row_ptrs[row]; k < row_ptrs[row + 1]; ++k) {
                 auto col = col_idxs[k];
                 if (col > row) {
-                    x->at(row, j) -= vals[k] * x->at(col, j);
+                    ValueType xval{};
+                    while (is_nan(xval = load_relaxed(x->at(col, j)))) {
+                    }
+                    val -= vals[k] * xval;
                 }
                 if (col == row) {
                     diag = vals[k];
                 }
             }
             if (!unit_diag) {
-                x->at(row, j) /= diag;
+                val /= diag;
+            }
+            if (is_nan(val)) {
+                val = zero<ValueType>();
```

Comment:
For CUDA, we warn that a `NaN` was produced in the solve, for reference we just keep the `NaN`s in the solution if any are produced. I think we should warn here to, such as not to suggest we produced a correct solution if we didn't.